diff --git a/src/streamlink/plugins/twitch.py b/src/streamlink/plugins/twitch.py
index e74e901a..547ca55f 100644
--- a/src/streamlink/plugins/twitch.py
+++ b/src/streamlink/plugins/twitch.py
@@ -23,7 +23,7 @@ from datetime import datetime, timedelta
 from json import dumps as json_dumps
 from random import random
 from typing import ClassVar, Mapping, Optional, Tuple, Type
-from urllib.parse import urlparse
+from urllib.parse import quote, urlparse
 
 from requests.exceptions import HTTPError
 
@@ -44,7 +44,7 @@ from streamlink.stream.hls import (
     parse_tag,
 )
 from streamlink.stream.http import HTTPStream
-from streamlink.utils.args import keyvalue
+from streamlink.utils.args import comma_list, keyvalue
 from streamlink.utils.parse import parse_json, parse_qsd
 from streamlink.utils.random import CHOICES_ALPHA_NUM, random_token
 from streamlink.utils.times import fromtimestamp, hours_minutes_seconds_float
@@ -54,6 +54,7 @@ from streamlink.utils.url import update_qsd
 log = logging.getLogger(__name__)
 
 LOW_LATENCY_MAX_LIVE_EDGE = 2
+STREAMLINK_TTVLOL_VERSION = "173b8c93-master"
 
 
 @dataclass
@@ -245,6 +246,67 @@ class UsherService:
         return self._create_url(f"/vod/{video_id}", **extra_params)
 
 
+class NoPlaylistProxyAvailable(Exception):
+    """
+    No playlist proxies available.
+    """
+
+
+class PlaylistProxyService:
+    def __init__(self, session, playlist_proxies, excluded_channels, fallback):
+        self.session = session
+        self.playlist_proxies = playlist_proxies or []
+        self.excluded_channels = map(str.lower, excluded_channels or [])
+        self.fallback = fallback
+
+    def _append_query_params(self, url):
+        params = {
+            "player": "twitchweb",
+            "type": "any",
+            "allow_source": "true",
+            "allow_audio_only": "true",
+            "allow_spectre": "false",
+            "fast_bread": "true",
+        }
+        req = self.session.http.prepare_new_request(url=url, params=params)
+
+        return req.url
+
+    def streams(self, channel, **kwargs):
+        if not self.playlist_proxies:
+            raise NoPlaylistProxyAvailable
+
+        if channel in self.excluded_channels:
+            log.info(f"Channel {channel} excluded from playlist proxy")
+            raise NoPlaylistProxyAvailable
+
+        log.debug(f"Getting live HLS streams for {channel}")
+        self.session.http.headers.update({
+            "referer": "https://player.twitch.tv",
+            "origin": "https://player.twitch.tv",
+        })
+        for proxy in self.playlist_proxies:
+            url = re.sub(r"\[channel\]", channel, proxy)
+            parsed_url = urlparse(url)
+
+            if url == proxy:
+                url = quote(self._append_query_params(url + f"/playlist/{channel}.m3u8"), safe=":/")
+            elif not parsed_url.query:
+                url = self._append_query_params(url)
+
+            log.info(f"Using playlist proxy '{parsed_url.scheme}://{parsed_url.netloc}'")
+            try:
+                return TwitchHLSStream.parse_variant_playlist(self.session, url, **kwargs)
+            except OSError as err:
+                log.error(err)
+
+        if self.fallback:
+            log.info("No playlist proxies available, falling back to Twitch servers")
+            raise NoPlaylistProxyAvailable
+
+        raise NoStreamsError
+
+
 class TwitchAPI:
     CLIENT_ID = "kimne78kx3ncx6brgo4mv6wki5h1ko"
 
@@ -688,6 +750,59 @@ class TwitchClientIntegrity:
     action="store_true",
     help="Purge cached Twitch client-integrity token and acquire a new one.",
 )
+@pluginargument(
+    "proxy-playlist",
+    metavar="URLS",
+    type=comma_list,
+    help="""
+        Proxy the playlist request through a server specified at <URL>.
+
+        If the URL has no path the playlist will be requested using the TTVLOL API.
+        If the URL path includes [channel] the playlist will not be requested with the TTVLOL API and
+        [channel] will be replaced with the channel name at runtime.
+
+        Can be multiple comma separated server URLs to be used as fallback.
+
+        When used the Twitch GraphQL API will not be called.
+        Only livestreams will use the playlist proxy, VODs and clips will use upstream behavior.
+        Integrity token retrieval will not be attempted.
+        --twitch-api-header, --twitch-access-token-param, and --twitch-purge-client-integrity will have no effect.
+        It will also not be possible to check for subscriber only streams.
+    """,
+)
+@pluginargument(
+    "proxy-playlist-exclude",
+    metavar="CHANNELS",
+    type=comma_list,
+    help="""
+        Exclude specified channel(s) from playlist proxy and fallback to upstream behavior.
+
+        Can be multiple comma separated channel names.
+
+        Useful if you're subscribed to the channel(s) and want to use your OAuth token to avoid ads instead.
+    """,
+)
+@pluginargument(
+    "proxy-playlist-fallback",
+    action="store_true",
+    help="""
+        Fallback to Twitch servers if all requests to playlist proxies fail.
+    """,
+)
+@pluginargument(
+    "ttvlol",
+    action="store_true",
+    help="""
+        Deprecated and will be removed in a future release.
+    """,
+)
+@pluginargument(
+    "reexec-on-ad",
+    action="store_true",
+    help="""
+        Deprecated and will be removed in a future release.
+    """,
+)
 class Twitch(Plugin):
     _CACHE_KEY_CLIENT_INTEGRITY = "client-integrity"
 
@@ -708,6 +823,9 @@ class Twitch(Plugin):
         self.clip_name = None
         self._checked_metadata = False
 
+        log.info(f"streamlink-ttvlol {STREAMLINK_TTVLOL_VERSION} ({self.session.version})")
+        log.info("Please report issues to https://github.com/2bc4/streamlink-ttvlol/issues")
+
         if self.subdomain == "player":
             # pop-out player
             if self.params.get("video"):
@@ -727,6 +845,12 @@ class Twitch(Plugin):
             access_token_param=self.get_option("access-token-param"),
         )
         self.usher = UsherService(session=self.session)
+        self.playlist_proxy = PlaylistProxyService(
+                session=self.session,
+                playlist_proxies=self.get_option("proxy-playlist"),
+                excluded_channels=self.get_option("proxy-playlist-exclude"),
+                fallback=self.get_option("proxy-playlist-fallback"),
+        )
 
         def method_factory(parent_method):
             def inner():
@@ -901,7 +1025,14 @@ class Twitch(Plugin):
         elif self.clip_name:
             return self._get_clips()
         elif self.channel:
-            return self._get_hls_streams_live()
+            try:
+                return self.playlist_proxy.streams(
+                    channel=self.channel,
+                    disable_ads=self.get_option("disable-ads"),
+                    low_latency=self.get_option("low-latency"),
+                )
+            except NoPlaylistProxyAvailable:
+                return self._get_hls_streams_live()
 
 
 __plugin__ = Twitch
